### PR TITLE
Code review request

### PR DESCRIPTION
**Pull request by zdrale**
-In this project, we have implemented our own version of Ruby’s enumerable methods.
-These are the methods we recreated:
my_each
my_each_with_index
my_select
my_all?
my_any?
my_none?
my_count
my_map
my_inject
multiply_els

### CODE REVIEW

File: methods.rb [hunk 1]
```diff
@@ -0,0 +1,112 @@
+module Enumerable
+  def my_each
+    for i in self
+      yield i
+    end
+  end
```
Comment (arslanbisharat) [inline methods.rb]: * [ ] should have working `my_each`  identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_each
- [ ] my_each returns an Enumerator if no block is given

**- Yours is always expecting a block**



**- Real each**
image

File: methods.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+  def my_each_with_index
+    i = 0
+    my_each do |item|
+      yield item, i
+      i += 1
+    end
+  end
```
Comment (arslanbisharat) [inline methods.rb]: * [ ] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_each_with_index

- [ ] my_each_with_index returns an Enumerator if no block is given

- Yours is always expecting a block



- Real each_with_index
image

File: methods.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  def my_select
+    arry = []
+    for i in self
+      arry << i if yield i
+    end
+    arry
+  end
```
Comment (arslanbisharat) [inline methods.rb]: * [ ] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_select
- [ ] my_select returns an enumerator if no block is given

**- Yours is always expecting a block**



**- Real select**
image

File: methods.rb [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    for i in self
+      if initialize_num
+        sum = i
+        initialize_num = false
+      else
+        sum = yield sum, i
+      end
+    end
+    sum
+  end
```
Comment (arslanbisharat) [inline methods.rb]: - [ ] should have working `my_inject` **(covered by automated tests)** identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```

- Real inject
image

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13
```
- Yours is always expecting a block



**- Real inject**
image

File: methods.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                     end
+    output = true if filtered_array.to_a.empty?
+    output
+  end
+
+  def my_count
+    count = 0
+    count += 1 if size > 1
+    yield count
+  end
```
Comment (arslanbisharat) [inline methods.rb]: * [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_count

- [ ] my_count returns the number of items in enum through enumeration
```ruby
[5, 4, 1].my_count == [5, 4, 1].count => 3
```
- Yours is always expecting a block



- Real count
image

- [ ] my_count counts the number of items in enum that are equal to the item if an argument is given
```ruby
[5, 4, 1].my_count(4) == [5, 4, 1].count(4) => 1
```
- Yours is giving an ArgumentError


- Real count
image

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
# Request changes 😳

I am very happy to see this project and I can see you worked hard on this project. There are only a few issues that we need to address and I am sure that you will end up with a perfect project

_It's okay. We are here to learn. Don't get discouraged. Work even harder. You are almost there.😄_

**Please have a look at the suggestions to get more understanding.**

- [ ] Please remove linter erros. 

- [x] Please add a proper description in your readme file.

- [x] You have 10 failing tests. I have given suggestions below. Please have a look at them and solve them. A potential source for learning is [this](https://www.rubyguides.com/2016/03/enumerable-methods/)

**Please make the changes and request re-review for the given project**
Stay safe and give me a smile 😇

### Happy Coding
---------------------------------------------------------------------------------------------------------------------------------
**Regards,
Arslan Bisharat
slack: arslanbisharat**

**Comment by mimipeshy** [comment]
## STATUS :CHANGES REQUIRED :red_circle: 

Hey guys :wave:  :wave: , great job on working on most of the feedback given by the previous TSE, however we still have a few more changes to work on :smile: 

- [ ] Work on the linter errors you are having


Check the comments for more details on the same

Regards
Peris Ndanu

File: methods.rb [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                     elsif arg.is_a?(Class)
+                       my_select { |el| el.class <= arg }
+                     else
+                       my_select { |el| el == arg }
+                     end
+    output = true if filtered_array.to_a.empty?
+    output
+  end
+
+  def my_count (argument=nil)
```
Comment (mimipeshy) [inline methods.rb:72]: - [ ] If a block is given, it counts the number of elements yielding a true value.
true

- What it expects `array.count{ |x| x%2==0 } #=> 3`

- [ ] it should not change the original array
mutate

Comment (mimipeshy) [inline methods.rb:72]: #my_count
- [ ] If a block is given, it counts the number of elements yielding a true value.
true

- What it expects `array.count{ |x| x%2==0 } #=> 3`

- [ ] it should not change the original array
mutate

File: methods.rb [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    if proc_block.class == Proc and block_given?
+      my_each { |el| new_arr.push(proc_block.call(el)) }
+    else
+      my_each { |el| new_arr.push(yield(el)) }
+    end
+
+    new_arr
+  end
+
+  def my_inject(init = nil, arg = nil)
```
Comment (mimipeshy) [inline methods.rb:99]: - [ ] enumerables #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
inject

- [ ]  enumerables #my_inject when a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method: range

range

File: methods.rb [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        sum = i
+        initialize_num = false
+      else
+        sum = yield sum, i
+      end
+    end
+    sum
+  end
+
+  def multiply_els(arr)
```
Comment (mimipeshy) [inline methods.rb:119]: - [ ] enumerables #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
els

File: methods.rb [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+   counter = 0
+   if block_given?
+    to.a.my_each {|item| counter += 1 if yield(item)}
+   elsif !argument.nil?
+    to_a.my_each{|item| counter += 1 if argument == item}
+  else
+    counter = to_a.length
+  end
+  counter
+  end
```
Comment (mimipeshy) [inline methods.rb]: #my_count
- [ ] If a block is given, it counts the number of elements yielding a true value.
true

- What it expects `array.count{ |x| x%2==0 } #=> 3`

- [ ] it should not change the original array
mutate

File: methods.rb [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    for i in self
+      if initialize_num
+        sum = i
+        initialize_num = false
+      else
+        sum = yield sum, i
+      end
+    end
+    sum
+  end
```
Comment (mimipeshy) [inline methods.rb:118]: - [ ] enumerables #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method

**Yours is always expecting a block**
inject


- [ ]  enumerables #my_inject when a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method: range

**Yours is always expecting a block**
range

File: methods.rb [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+  def multiply_els(arr)
+    arr.my_inject { |n, total| n * total }
+  end
+
+end
```
Comment (mimipeshy) [inline methods.rb:123]: - [ ] enumerables #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
els

File: methods.rb [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        sum = i
+        initialize_num = false
+      else
+        sum = yield sum, i
+      end
+    end
+    sum
+  end
+
+  def multiply_els(arr)
```
Comment (layzrd) [inline methods.rb:120]: Move the #multiply_els method out of the module.

File: methods.rb [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    if proc_block.class == Proc and block_given?
+      my_each { |el| new_arr.push(proc_block.call(el)) }
+    else
+      my_each { |el| new_arr.push(yield(el)) }
+    end
+
+    new_arr
+  end
+
+  def my_inject(init = nil, arg = nil)
```
Comment (layzrd) [inline methods.rb]: #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range.
Fix the method to make it work for a symbol.

File: methods.rb [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+                       my_select { |el| el.class <= arg }
+                     else
+                       my_select { |el| el == arg }
+                     end
+    output = true if filtered_array.to_a.empty?
+    output
+  end
+
+  def my_count (argument=nil)
+    if num
```
Comment (layzrd) [inline methods.rb]: undefined local variable or method `num'. fix it.

**Review by layzrd — CHANGES_REQUESTED** [review]
### Status: Require changes  ♲
Hi @zdrale  👏 You did very well on this project. ✨ I applaud for successfully implementing custom enumerable methods. Most requirements have done 👌, Kindly fix the following issues.

[REQUIRED]

- [ ] Fix syntax  errors that I comment inline the code
- [ ] Fix the my_inject method to enable it to accept symbols. You have to add all features to make your methods work like the original ruby enumerator. Check the following link https://ruby-doc.org/core-2.7.2/Enumerable.html.
- [ ] Fix the rubocop linting errors.

Once you fix the issues, kindly submit another code review. ♲

Happy coding 👨‍💻
Abdulaziz Redi Ali

**Review by jubaan — APPROVED** [review]
# :tada: :tada: :tada: :rotating_light: Status: Approved :rotating_light:  :tada: :tada: :tada:

#### :clap: :clap: Great job :hammer_and_pick:!! There are no more issues left to be addressed. 

- [x] Please fix all linter errors before merging to master.

## See you at the top :rocket:!

# :tada: :tada: :tada: :rotating_light: Status: Approved :rotating_light:  :tada: :tada: :tada:

**Pull request merged**